### PR TITLE
[FLINK-31646][network] Implement the remote tier producer for the tiered storage

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

*Implement the remote tier producer for the tiered storage*

## Brief change log

*(for example:)*
  - *【commit 1】 Implement HashPartitionFile for the tiered storage*
  - *【commit 2】Implement the remote tier producer for the tiered storage*

## Verifying this change

This change will add tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    private void releaseAllResources() {
+        cacheDataManager.release();
+    }
+
+    private void emitBuffer(Buffer finishedBuffer, int subpartitionId) {
+        cacheDataManager.appendBuffer(finishedBuffer, subpartitionId);
+    }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:109]: 1. The name `emit` is confusing.
2. Why do we need this method when its body is only one line of code?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+/** The implementation of {@link TierProducerAgent} for the remote tier. */
+public class RemoteTierProducerAgent implements TierProducerAgent {
+
+    private final int numSubpartitions;
+
+    private final int numBuffersPerSegment;
+
+    private final RemoteCacheManager cacheDataManager;
+
+    private final int[] currentSubpartitionWriteBuffers;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:42]:
```suggestion
    private final int[] currentSubpartitionSegmentWriteBuffers;
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        cacheDataManager.startSegment(subpartitionId.getSubpartitionId(), segmentId);
+        // The remote storage tier should always be able to start a new segment.
+        return true;
+    }
+
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer buffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (currentSubpartitionWriteBuffers[subpartitionIndex] != 0
+                && currentSubpartitionWriteBuffers[subpartitionIndex] + 1 > numBuffersPerSegment) {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:81]: We already checked in the constructor that one segment at least contain one buffer. `currentSubpartitionWriteBuffers[subpartitionIndex] != 0` is probably not needed here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        storageMemoryManager.listenBufferReclaimRequest(this::flushBuffers);
+    }
+
+    // ------------------------------------------------------------------------
+    //  Called by RemoteCacheManager
+    // ------------------------------------------------------------------------
+
+    void startSegment(int segmentId) {
+        synchronized (allBuffers) {
+            this.segmentId = segmentId;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:102]: Should check `allBuffers` is empty.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                            subpartitionId,
+                            Collections.singletonList(
+                                    new PartitionFileWriter.SegmentBufferContext(
+                                            segmentId, Collections.emptyList(), true)));
+            // Notify the partition file writer that the segment is finished through writing the
+            // buffer context
+            flushCompletableFuture =
+                    partitionFileWriter.write(
+                            partitionId, Collections.singletonList(bufferContext));
+        }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:130]: This doesn't seem right. `numFlushedBuffers == 0` doesn't mean there's no data in the segment. It is possible that the buffers are flushed before calling `finishSegment`. E.g., on buffer reclaim request.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java [hunk 6]
```diff
@@ -49,4 +60,61 @@ public static void setBufferWithHeader(
...
+            String basePath, TieredStoragePartitionId partitionId, int subpartitionId) {
+        while (basePath.endsWith("/") && basePath.length() > 1) {
+            basePath = basePath.substring(0, basePath.length() - 1);
+        }
+        return String.format(
+                "%s/%s/%s",
+                basePath, TieredStorageIdMappingUtils.convertId(partitionId), subpartitionId);
+    }
+
+    public static Path getSegmentPath(
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java:108]: `setBufferWithHeader` might be private.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/TierFactory.java [hunk 7]
```diff
@@ -32,7 +32,9 @@
 public interface TierFactory {
 
     /** Creates the master-side agent of a Tier. */
-    TierMasterAgent createMasterAgent(TieredStorageResourceRegistry tieredStorageResourceRegistry);
+    TierMasterAgent createMasterAgent(
+            String remoteStorageBaseHomePath,
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/TierFactory.java:36]: I think `dataFileBasePath`, `partitionFileWriter` and `partitionFileReader` shares the same problem.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java]: 1. Moved the methods about `segment mode` to `SegmentPartitionFile`.
2. Moved `DATA_FILE_SUFFIX` to `ProducerMergedPartitionFile`
3. Make `setBufferWithHeader` private.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFile.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.file;
+
+/**
+ * The partition file in the hash mode. In this mode, each segment of one subpartition is written to
+ * an independent file.
+ */
+public class HashPartitionFile {
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFile.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * The implementation of {@link PartitionFileWriter} with hash mode. In this mode, each segment of
+ * one subpartition is written to an independent file.
+ *
+ * <p>Note that after finishing writing a segment, a segment-finish file is written to ensure the
+ * downstream reads only when the entire segment file is written, avoiding partial data reads. The
+ * downstream can determine if the current segment is complete by checking for the existence of the
+ * segment-finish file.
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java]: Fixed. Each subpartition only keeps a segment-finish file for the latest segment.
Added doc in `SegmentPartitionFileWriter`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+                            segmentBuffers -> {
+                                CompletableFuture<Void> flushSuccessNotifier =
+                                        new CompletableFuture<>();
+                                Runnable writeRunnable =
+                                        getFlushOrFinisheSegmentRunnable(
+                                                partitionId,
+                                                subpartitionId,
+                                                segmentBuffers,
+                                                flushSuccessNotifier);
+                                ioExecutor.execute(writeRunnable);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+                subpartitionBuffers -> {
+                    int subpartitionId = subpartitionBuffers.getSubpartitionId();
+                    List<SegmentBufferContext> multiSegmentBuffers =
+                            subpartitionBuffers.getSegmentBufferContexts();
+                    multiSegmentBuffers.forEach(
+                            segmentBuffers -> {
+                                CompletableFuture<Void> flushSuccessNotifier =
+                                        new CompletableFuture<>();
+                                Runnable writeRunnable =
+                                        getFlushOrFinisheSegmentRunnable(
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+                ? () ->
+                        flush(
+                                partitionId,
+                                subpartitionId,
+                                segmentId,
+                                buffersToFlush,
+                                flushSuccessNotifier)
+                : () ->
+                        writeFinishSegmentFile(
+                                partitionId, subpartitionId, segmentId, flushSuccessNotifier);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java]: 1. Fixed
2. `DiskCacheManager#getSubpartitionToFlushBuffers` will be fixed in the subsequential change, which is not fixed in this PR.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/TierFactory.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/TierFactory.java]: 1. Moved `remoteStorageBasePath` to the constructor.
2. `dataFileBasePath` is the result partition path, which should be an argument.
3. `partitionFileWriter` and `partitionFileReader` has been removed from the arguments. They are initialized by calling the static methods of the partition file.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierMasterAgent.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            LOG.error("Failed to release the partition file for {}", partitionID);
+        }
+
+        resourceRegistry.clearResourceFor(partitionID);
+    }
+
+    @Override
+    public void release(String toRelease) {
+        deletePathQuietly(toRelease);
+    }
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierMasterAgent.java]: Removed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java (hunk 1 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java (hunk 2 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java]: Renamed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java (hunk 3 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java (hunk 4 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:103]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java (hunk 5 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java]: Fixed.

**Comment by TanYuxin-tyx** [comment]
@xintongsong Thanks for helping review. I addressed the comments with a fix-up commit and added the unit tests.
But the fix-up is only to be convenient for review. The fixup's files are not separated correctly. I will rebase and separate the files again after the review.

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Review by reswqa — COMMENTED** [review]
Thanks @TanYuxin-tyx, I left some comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFile.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        FileStatus[] files = fs.listStatus(segmentFinishDir);
+        if (files.length == 0) {
+            OutputStream outputStream =
+                    fs.create(segmentFinishFile, FileSystem.WriteMode.OVERWRITE);
+            outputStream.close();
+        } else {
+            // Note that this check requires the file system to ensure that only one file is in the
+            // directory can be accessed when renaming a file.
+            checkState(files.length == 1, "Wong number of segment-finish files.");
+            fs.rename(files[0].getPath(), segmentFinishFile);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFile.java:143]: Added doc to explain this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+                    partitionFileWriter.write(
+                            partitionId, Collections.singletonList(subpartitionBufferContext));
+        }
+    }
+
+    private void recycleBuffers() {
+        synchronized (allBuffers) {
+            for (Tuple2<Buffer, Integer> bufferAndIndex : allBuffers) {
+                Buffer buffer = bufferAndIndex.f0;
+                if (!buffer.isRecycled()) {
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java]: Ok, simplified this method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+    void addBuffer(Buffer buffer) {
+        Tuple2<Buffer, Integer> toAddBuffer = new Tuple2<>(buffer, bufferIndex++);
+        synchronized (allBuffers) {
+            allBuffers.add(toAddBuffer);
+        }
+    }
+
+    void finishSegment(int segmentId) {
+        synchronized (allBuffers) {
+            checkState(this.segmentId == segmentId, "Wrong segment id.");
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java]: Removed the synchronization. Only the task thread can modify the segmentId, and the method can only be called by the task thread, so accessing segmentId is not guarded here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFileWriter.java [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+
+    private void flushOrFinishSegment(
+            TieredStoragePartitionId partitionId,
+            int subpartitionId,
+            SegmentBufferContext segmentBuffers,
+            CompletableFuture<Void> flushSuccessNotifier) {
+        int segmentId = segmentBuffers.getSegmentId();
+        List<Tuple2<Buffer, Integer>> buffersToFlush = segmentBuffers.getBufferAndIndexes();
+        boolean isSegmentFinished = segmentBuffers.isSegmentFinished();
+        checkState(!buffersToFlush.isEmpty() || isSegmentFinished);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFileWriter.java:141]: This is not exclusive, so this is not changed.

**Comment by TanYuxin-tyx** [comment]
@reswqa Thanks for helping review, I have addressed the comments, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+            LOG.error("Failed to flush the buffers.", e);
+            ExceptionUtils.rethrow(e);
+        }
+        flushBuffers();
+    }
+
+    /** Release all buffers. */
+    void release() {
+        recycleBuffers();
+        checkState(allBuffers.isEmpty(), "Leaking buffers.");
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java]: Added the synchronization to protect this.

**Review by reswqa — APPROVED** [review]
Thanks for the update, LGTM.

**Comment by TanYuxin-tyx** [comment]
@xintongsong @reswqa Thanks for helping review.

**Pull request merged**